core: combined inode_link and inode_lookup

rkothiya opened the pull request:
combined inode_link and inode_lookup into more efficient function
inode_link_lookup where we can avoid reference and unreference of
the linked_inode.

Fixes: #1341

Change-Id: Ic878b2c028a422a359aa45d7dd0824faa0e35c05
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

Commit:
core: combined inode_link and inode_lookup

combined inode_link and inode_lookup into more efficient function
inode_link_lookup where we can avoid reference and unreference of
the linked_inode.

Fixes: #1341

Change-Id: Ic878b2c028a422a359aa45d7dd0824faa0e35c05
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

pranithk commented:
@rkothiya Could you attach any results of the benchmarks where performance improved because of this patch?

pranithk reviewed (changes requested):
Keeping it in request changes state until the data that proves that this patch improves performance on some benchmark is provided.

mykaul commented on libglusterfs/src/inode.c:
```diff
@@ -1102,6 +1102,43 @@ inode_link(inode_t *inode, inode_t *parent, const char *name, struct iatt *iatt)
+        hash = hash_dentry(parent, name, table->dentry_hashsize);
+    }
+
+    if (name && strchr(name, '/')) {
```
nit: can move this check to line 1118

rkothiya replied:
Done

mykaul commented:
> Keeping it in request changes state until the data that proves that this patch improves performance on some benchmark is provided.

Makes sense, although I'd argue:
1. It improves readability of the code, IMHO.
2. It may be a step in the right direction. For example, a follow up patch can ensure inode_table_prune() already 'inherits' the lock from this code instead of the unlock, lock sequence.


@pranithk - do you have a specific performance test in mind?

mykaul commented:
Note: without combining the inode_unref() into this, we are not removing a lock of the table, which I assume is probably more contented and will provide more real value. Issue #1341 mentioned it explicitly.

pranithk commented:
> > Keeping it in request changes state until the data that proves that this patch improves performance on some benchmark is provided.
> 
> Makes sense, although I'd argue:

I am seeing a pattern of PRs which say they are improving performance without any data showing that it indeed did improve performance.

> 
> 1. It improves readability of the code, IMHO.
> 2. It may be a step in the right direction. For example, a follow up patch can ensure inode_table_prune() already 'inherits' the lock from this code instead of the unlock, lock sequence.

I don't disagree. My only point is that it is important that we **know** that a particular change is improving performance and by how much percentage.

> 
> @pranithk - do you have a specific performance test in mind?

Any benchmark that you think would improve performance is fine. Just post the numbers before and after the change. If it is scalability improvement, give numbers about what was the scalability limit before and after the change.

A sample bug where such information is provided is: https://bugzilla.redhat.com/show_bug.cgi?id=1676479

pranithk commented:
> Note: without combining the inode_unref() into this, we are not removing a lock of the table, which I assume is probably more contented and will provide more real value. Issue #1341 mentioned it explicitly.

My only request is to quantify the 'real value' so that we all can discuss objectively about the improvement in performance/scalability that this change brought about.

mykaul commented:
> > > Keeping it in request changes state until the data that proves that this patch improves performance on some benchmark is provided.
> > 
> > 
> > Makes sense, although I'd argue:
> 
> I am seeing a pattern of PRs which say they are improving performance without any data showing that it indeed did improve performance.

I wish we could run performance tests for each PR, not only to asses performance but also to ensure patches do not introduce regressions in performance.

> 
> > 1. It improves readability of the code, IMHO.
> > 2. It may be a step in the right direction. For example, a follow up patch can ensure inode_table_prune() already 'inherits' the lock from this code instead of the unlock, lock sequence.
> 
> I don't disagree. My only point is that it is important that we **know** that a particular change is improving performance and by how much percentage.
> 
> > @pranithk - do you have a specific performance test in mind?
> 
> Any benchmark that you think would improve performance is fine. Just post the numbers before and after the change. If it is scalability improvement, give numbers about what was the scalability limit before and after the change.
> 
> A sample bug where such information is provided is: https://bugzilla.redhat.com/show_bug.cgi?id=1676479

Let's run the whole performance suite then. I'd argue that we also need to run some performance tests with many files.
From looking at the code, it seems be called on mkdir, mknod, create, symlink.

xhernandez commented on xlators/protocol/server/src/server-rpc-fops_v2.c:
```diff
@@ -2109,16 +2108,7 @@ server4_icreate_cbk(call_frame_t *frame, void *cookie, xlator_t *this,
-
-    inode_lookup(link_inode);
-    inode_unref(link_inode);
+    inode_link_lookup(inode, state->loc.parent, state->loc.name, stbuf);
```
Based on old code, this could potentially fail. Shouldn't we return errors from `inode_link_lookup()` ?

rkothiya replied:
Done

xhernandez commented on libglusterfs/src/inode.c:
```diff
@@ -1102,6 +1102,43 @@ inode_link(inode_t *inode, inode_t *parent, const char *name, struct iatt *iatt)
+        return;
+    }
+
+    pthread_mutex_lock(&table->lock);
+    linked_inode = __inode_link(inode, parent, name, iatt, hash);
+    pthread_mutex_unlock(&table->lock);
```
This introduces a race. We leave the mutex without having increased the reference count of `linked_inode`. This means that the inode could be unreferenced and destroyed after this point.

rkothiya replied:
I have made the changes to increase the reference count.

xhernandez commented on libglusterfs/src/inode.c:
```diff
@@ -1102,6 +1102,43 @@ inode_link(inode_t *inode, inode_t *parent, const char *name, struct iatt *iatt)
 }
 
+void
+inode_link_lookup(inode_t *inode, inode_t *parent, const char *name,
+                  struct iatt *iatt)
```
This is mostly the same as `inode_link()`. Probably it would be better to move the common code to a new function instead of duplicating it.

rkothiya replied:
Combined the inode_link and inode_lookup in this function as suggested in the opened issue.

Commit:
core: combined inode_link and inode_lookup

As per the review comments added the code to increase
the reference count and unreference the inode to avoid
race condition.

Fixes: #1341

Change-Id: If2fb6744586b6bdd73bcbdde73d69ad4153c1ae0
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

rkothiya commented:
/recheck smoke

rkothiya commented:
As per the request I have generated the perf report and got the below output with and without the patch. 
But I do not think that the impact of these changes would be visible in the below report. 
~~~
===========================================================================================================================
FOPs     glusterfs-10dev-0.37.git738b23b.el7   glusterfs-10dev-0.39.git11350c7.el7     glusterfs-10dev-0.39.git11350c7.el7
                   (without patch)                    (with the current patch)                          vs
                                                                                       glusterfs-10dev-0.37.git738b23b.el7
===========================================================================================================================
create                           927                            914                              -1
ls-l                             14127                          14403                             1
chmod                            1002                           964                              -3
stat                             1139                           1098                             -3
read                             1170                           1170                              0
append                           739                            738                               0
rename                           478                            490                               2
delete-renamed                   5378                           5499                              2
mkdir                            3238                           3281                              1
rmdir                            2766                           2734                             -1
cleanup                          7498                           7495                              0
===========================================================================================================================
~~~

pranithk commented:
> As per the request I have generated the perf report and got the below output with and without the patch.
> But I do not think that the impact of these changes would be visible in the below report.
> 
> ```
> ===========================================================================================================================
> FOPs     glusterfs-10dev-0.37.git738b23b.el7   glusterfs-10dev-0.39.git11350c7.el7     glusterfs-10dev-0.39.git11350c7.el7
>                    (without patch)                    (with the current patch)                          vs
>                                                                                        glusterfs-10dev-0.37.git738b23b.el7
> ===========================================================================================================================
> create                           927                            914                              -1
> ls-l                             14127                          14403                             1
> chmod                            1002                           964                              -3
> stat                             1139                           1098                             -3
> read                             1170                           1170                              0
> append                           739                            738                               0
> rename                           478                            490                               2
> delete-renamed                   5378                           5499                              2
> mkdir                            3238                           3281                              1
> rmdir                            2766                           2734                             -1
> cleanup                          7498                           7495                              0
> ===========================================================================================================================
> ```

This is what I was afraid of. I think bottlenecks in glusterfs are in the fs, network layer. i.e. number of rpcs, number of fs-syscalls or synchronization during these fs/network layer operations. So patches like this may not have any perceivable effect. In my book, the effort is better spent elsewhere. Check this link https://gist.github.com/jboner/2841832 - for comparison of latency between memory ops vs disk ops

rkothiya commented:
I agree, that we are not seeing any improvement or negative effect of this patch. 
We definitely should focus on areas which can cause a bigger impact.
IMHO these patches can take a low priority but they would make our code more efficient.
